File: PatentReverseCiting/PatentReverseCiting.py
from pyspark import SparkContext, SparkConf
import sys

inputs = sys.argv[1]
outputs = sys.argv[2]

conf = SparkConf().setAppName("patent reverse citing")
conf.set("spark.hadoop.validateOutputSpecs", "false")
context = SparkContext(conf=conf)

rdd = context.textFile(inputs)\
			 .filter(lambda x:x!='"CITING","CITED"')\
			 .map(lambda x:x.split(','))\
			 .map(lambda x:(x[1].encode("utf-8"), x[0].encode("utf-8")))\
			 .reduceByKey(lambda x,y: x+','+y)\
			 .coalesce(1)\
			 .saveAsTextFile(outputs)

File: Scala/PatentReverseCiting/PatentReverseCiting.scala
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext

object PatentReverseCiting	{
	def main(args: Array[String])	{
		val inputFile = args(0)
		val outputFile = args(1)
		val conf = new SparkConf().setAppName("patent reverse citing")
		val sc = new SparkContext(conf)

		val inputData = sc.textFile(inputFile).filter(x => x != "\"CITING\",\"CITED\"")
		val rdd = inputData.map(line => line.split(","))
		val rddReverseTuple = rdd.map(x => (x(1), x(0)))
		val result = rddReverseTuple.reduceByKey((cite1, cite2) => cite1 + ',' + cite2)

		result.saveAsTextFile(outputFile)
	}
}